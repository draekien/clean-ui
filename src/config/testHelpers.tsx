import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
