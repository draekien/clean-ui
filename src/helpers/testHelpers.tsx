import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
