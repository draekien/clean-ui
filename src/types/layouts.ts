export type HorizontalAlignment = 'left' | 'center' | 'right';
export type VerticalAlignment = 'top' | 'center' | 'bottom';
export type Direction = 'row' | 'column';
export type Size = 'small' | 'medium' | 'large';
