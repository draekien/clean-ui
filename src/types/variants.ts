export type StatusVariant = 'success' | 'info' | 'error' | 'warning';
