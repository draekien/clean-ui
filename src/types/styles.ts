import { Theme } from '@theme-ui/css';
import { colors } from '../components/theme/theme.colors';
import { layouts } from '../components/theme/theme.layouts';

/**
 * The colors defined in the theme.
 * @remarks
 * See {@link colors} for more details.
 */
export type Color = keyof typeof colors;
/** The color computed from the theme. */
export type ComputedColor = (t: Theme) => string;
/**
 * The spacings defined in the theme.
 * @remarks
 * See {@link layouts.space} for more details.
 */
export type Spacing = keyof typeof layouts.space;
/**
 * The radii defined in the theme.
 * @remarks
 * See {@link layouts.radii} for more details.
 */
export type Radii = keyof typeof layouts.radii;
/**
 * The shadows defined in the theme.
 * @remarks
 * See {@link layouts.shadows} for more details.
 */
export type Shadows = keyof typeof layouts.shadows;
/**
 * The zIndices defined in the theme.
 * @remarks
 * See {@link layouts.zIndices} for more details.
 */
export type ZIndex = keyof typeof layouts.zIndices;
