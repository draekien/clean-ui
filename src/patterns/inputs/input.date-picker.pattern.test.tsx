import React from 'react';
import { DatePickerInput } from './input.date-picker.pattern';
import { themedRender } from '../../helpers/testHelpers';
import { findByTestId, findByText, fireEvent } from '@testing-library/dom';
import { HorizontalAlignment } from '../../types/layouts';

describe('Datepicker pattern', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it('should match snapshot', () => {
    const { container } = themedRender(
      <DatePickerInput inputId="picker" label="datepicker" />
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when all props are populated', () => {
    const handleChange = jest.fn();
    const handleFocus = jest.fn();
    const handleBlur = jest.fn();
    const { container } = themedRender(
      <DatePickerInput
        inputId="picker"
        label="datepicker"
        helpText="datepicker help"
        startOfWeek="Monday"
        initialMonth={new Date(2020, 1, 1)}
        calendarPosition="center"
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        value="01/02/2020"
      />
    );

    const input = container.querySelector('input') as Element;
    fireEvent.change(input, {
      target: { value: '02/02/2020' },
    });

    fireEvent.focus(input, { bubbles: true });
    fireEvent.blur(input);

    expect(container).toMatchSnapshot();
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('should open calendar when button is clicked', async () => {
    const { container } = themedRender(
      <DatePickerInput
        inputId="picker"
        label="picker"
        selectedDate={new Date(2020, 1, 1)}
      />
    );

    const button = container.querySelector('button') as Element;

    fireEvent.click(button);

    const calendarLabel = await findByText(container, 'February 2020');
    expect(calendarLabel).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  ['left' as const, 'right' as const, 'center' as const].forEach(
    (position: HorizontalAlignment) => {
      it(`should open the calendar in the correct position when aligned ${position}`, async () => {
        const { container } = themedRender(
          <DatePickerInput
            inputId={position}
            label={position}
            calendarPosition={position}
          />
        );

        const button = container.querySelector('button') as Element;

        fireEvent.click(button);

        const calendar = await findByTestId(container, 'calendar-container');
        const styles = window.getComputedStyle(calendar);

        if (position === 'left') {
          expect(styles.left).toBe('0px');
        }

        if (position === 'right') {
          expect(styles.right).toBe('0px');
        }

        if (position === 'center') {
          expect(styles.left).toBe('calc(50% - 10em)');
        }
      });
    }
  );
});
