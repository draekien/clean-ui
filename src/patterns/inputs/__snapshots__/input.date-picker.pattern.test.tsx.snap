// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker pattern should match snapshot 1`] = `
<div>
  <div>
    <div
      class="css-1m0srfo-DatePickerInput"
    >
      <div
        class="css-109th3q-InputWrapper"
      >
        <div
          class="css-1qm0f6g-InputLabel"
        >
          <label
            class="css-1pst1pz-Label"
            for="picker"
          >
            datepicker
          </label>
        </div>
        <div
          class="css-xcsx4z"
        >
          <input
            class="css-msl6v2"
            id="picker"
            type="text"
            value=""
          />
          <div
            class="css-e7011r"
          >
            <button
              class="css-nnuopu-Button"
            >
              <span
                class="css-1geq0g0"
              >
                <i
                  class="material-icons  css-zif8ik-Icon"
                >
                  calendar_today
                </i>
              </span>
              <span />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Datepicker pattern should match snapshot when all props are populated 1`] = `
<div>
  <div>
    <div
      class="css-1m0srfo-DatePickerInput"
    >
      <div
        class="css-109th3q-InputWrapper"
      >
        <div
          class="css-1qm0f6g-InputLabel"
        >
          <label
            class="css-1pst1pz-Label"
            for="picker"
          >
            datepicker
          </label>
        </div>
        <div
          class="css-xcsx4z"
        >
          <input
            class="css-msl6v2"
            id="picker"
            type="text"
            value="02/02/2020"
          />
          <div
            class="css-e7011r"
          >
            <button
              class="css-nnuopu-Button"
            >
              <span
                class="css-1geq0g0"
              >
                <i
                  class="material-icons  css-zif8ik-Icon"
                >
                  calendar_today
                </i>
              </span>
              <span />
            </button>
          </div>
        </div>
        <small
          class="css-ftpxt1-InputWrapper"
        >
          datepicker help
        </small>
      </div>
    </div>
  </div>
</div>
`;

exports[`Datepicker pattern should open calendar when button is clicked 1`] = `
<div>
  <div>
    <div
      class="css-1m0srfo-DatePickerInput"
    >
      <div
        class="css-109th3q-InputWrapper"
      >
        <div
          class="css-1qm0f6g-InputLabel"
        >
          <label
            class="css-1pst1pz-Label"
            for="picker"
          >
            picker
          </label>
        </div>
        <div
          class="css-xcsx4z"
        >
          <input
            class="css-msl6v2"
            id="picker"
            type="text"
            value="01/02/2020"
          />
          <div
            class="css-e7011r"
          >
            <button
              class="css-nnuopu-Button"
            >
              <span
                class="css-1geq0g0"
              >
                <i
                  class="material-icons  css-zif8ik-Icon"
                >
                  calendar_today
                </i>
              </span>
              <span />
            </button>
          </div>
        </div>
      </div>
      <div
        class="css-1c8vey"
        data-testid="calendar-container"
      >
        <div
          class="css-r09zb2-Card"
        >
          <div>
            <div
              class="css-135ho84"
            >
              <div
                class="css-1pvc4u8-Calendar"
              >
                <div
                  class="css-ft1sxc-Calendar css-4cffwv"
                >
                  <button
                    class="css-hcal7n-Button"
                  >
                    <span
                      class="css-1geq0g0"
                    >
                      <i
                        class="material-icons  css-y0j37o-Icon"
                      >
                        chevron_left
                      </i>
                    </span>
                    <span />
                  </button>
                  <span
                    class="css-1a9mqxr-Text css-vurnku"
                  >
                    February
                     
                    2020
                  </span>
                  <button
                    class="css-hcal7n-Button"
                  >
                    <span />
                    <span
                      class="css-1geq0g0"
                    >
                      <i
                        class="material-icons  css-y0j37o-Icon"
                      >
                        chevron_right
                      </i>
                    </span>
                  </button>
                </div>
                <div
                  class="css-1a0kgyi-Calendar"
                >
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      S
                    </span>
                  </div>
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      M
                    </span>
                  </div>
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      T
                    </span>
                  </div>
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      W
                    </span>
                  </div>
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      T
                    </span>
                  </div>
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      F
                    </span>
                  </div>
                  <div
                    class="css-108m3ho-Tooltip"
                  >
                    <span
                      class="css-1rgqr4q"
                    >
                      S
                    </span>
                  </div>
                </div>
                <div
                  class="css-1tme9z0-Calendar"
                >
                  <span
                    class="css-gakn16-CalendarDay"
                    data-testid="Sun Jan 26 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-0"
                    id="Sun Jan 26 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-0"
                  >
                    26
                  </span>
                  <span
                    class="css-gakn16-CalendarDay"
                    data-testid="Mon Jan 27 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-1"
                    id="Mon Jan 27 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-1"
                  >
                    27
                  </span>
                  <span
                    class="css-gakn16-CalendarDay"
                    data-testid="Tue Jan 28 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-2"
                    id="Tue Jan 28 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-2"
                  >
                    28
                  </span>
                  <span
                    class="css-gakn16-CalendarDay"
                    data-testid="Wed Jan 29 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-3"
                    id="Wed Jan 29 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-3"
                  >
                    29
                  </span>
                  <span
                    class="css-gakn16-CalendarDay"
                    data-testid="Thu Jan 30 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-4"
                    id="Thu Jan 30 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-4"
                  >
                    30
                  </span>
                  <span
                    class="css-gakn16-CalendarDay"
                    data-testid="Fri Jan 31 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-5"
                    id="Fri Jan 31 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-5"
                  >
                    31
                  </span>
                  <span
                    class="css-1and8og-CalendarDay"
                    data-testid="Sat Feb 01 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-6"
                    id="Sat Feb 01 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-6"
                  >
                    1
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sun Feb 02 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-7"
                    id="Sun Feb 02 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-7"
                  >
                    2
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Mon Feb 03 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-8"
                    id="Mon Feb 03 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-8"
                  >
                    3
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Tue Feb 04 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-9"
                    id="Tue Feb 04 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-9"
                  >
                    4
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Wed Feb 05 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-10"
                    id="Wed Feb 05 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-10"
                  >
                    5
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Thu Feb 06 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-11"
                    id="Thu Feb 06 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-11"
                  >
                    6
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Fri Feb 07 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-12"
                    id="Fri Feb 07 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-12"
                  >
                    7
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sat Feb 08 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-13"
                    id="Sat Feb 08 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-13"
                  >
                    8
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sun Feb 09 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-14"
                    id="Sun Feb 09 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-14"
                  >
                    9
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Mon Feb 10 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-15"
                    id="Mon Feb 10 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-15"
                  >
                    10
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Tue Feb 11 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-16"
                    id="Tue Feb 11 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-16"
                  >
                    11
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Wed Feb 12 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-17"
                    id="Wed Feb 12 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-17"
                  >
                    12
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Thu Feb 13 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-18"
                    id="Thu Feb 13 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-18"
                  >
                    13
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Fri Feb 14 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-19"
                    id="Fri Feb 14 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-19"
                  >
                    14
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sat Feb 15 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-20"
                    id="Sat Feb 15 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-20"
                  >
                    15
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sun Feb 16 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-21"
                    id="Sun Feb 16 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-21"
                  >
                    16
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Mon Feb 17 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-22"
                    id="Mon Feb 17 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-22"
                  >
                    17
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Tue Feb 18 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-23"
                    id="Tue Feb 18 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-23"
                  >
                    18
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Wed Feb 19 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-24"
                    id="Wed Feb 19 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-24"
                  >
                    19
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Thu Feb 20 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-25"
                    id="Thu Feb 20 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-25"
                  >
                    20
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Fri Feb 21 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-26"
                    id="Fri Feb 21 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-26"
                  >
                    21
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sat Feb 22 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-27"
                    id="Sat Feb 22 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-27"
                  >
                    22
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sun Feb 23 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-28"
                    id="Sun Feb 23 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-28"
                  >
                    23
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Mon Feb 24 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-29"
                    id="Mon Feb 24 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-29"
                  >
                    24
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Tue Feb 25 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-30"
                    id="Tue Feb 25 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-30"
                  >
                    25
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Wed Feb 26 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-31"
                    id="Wed Feb 26 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-31"
                  >
                    26
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Thu Feb 27 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-32"
                    id="Thu Feb 27 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-32"
                  >
                    27
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Fri Feb 28 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-33"
                    id="Fri Feb 28 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-33"
                  >
                    28
                  </span>
                  <span
                    class="css-5ocht1-CalendarDay"
                    data-testid="Sat Feb 29 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-34"
                    id="Sat Feb 29 2020 00:00:00 GMT+1100 (Australian Eastern Daylight Time)-34"
                  >
                    29
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
